src/error.rs

```diff
@@ -76,7 +24,57 @@ macro_rules! impl_pyerr {
     };
 }
 
-impl_pyerr!(ShapeError);
+/// Represents that dimensionalities of the given arrays don't match.
+#[derive(Debug)]
+pub struct DimensionalityError {
+    from: usize,
+    to: usize,
+}
+
+impl DimensionalityError {
+    pub(crate) fn new(from: usize, to: usize) -> Self {
+        Self { from, to }
+    }
+}
+
+impl fmt::Display for DimensionalityError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let Self { from, to } = self;
+        write!(f, "dimensionality mismatch:\n from={}, to={}", from, to)
+    }
+}
+
+impl_pyerr!(DimensionalityError);
+
+/// Represents that types of the given arrays don't match.
+#[derive(Debug)]
+pub struct TypeError {
+    from: String,
+    to: String,
```

Comment:
I personally like having `Option<String>` and unwrapping it when formatting, but it's too nitpicking and either is fine

Reply:
I think in cases like this, having `Option<String>` may indicate that there may be type information available (i.e., logically available) or not, whereas here we always have both from and to types.

So I'd rather leave it like this if it's ok.